### PR TITLE
CM-789: Tuple Aliases in C#

### PR DESCRIPTION
**Pull request by AditiSaxena0289**
Initial Check in for the **MINI ARTICLE - Tuple Aliases in C#**

Jira Link - https://content.code-maze.com/browse/CM-789

**To run the code:**
\CodeMazeGuides\csharp-intermediate-topics\TupleAliasInCSharp\tuplealiasincsharp> **dotnet run**

**To run the tests:**
\CodeMazeGuides\csharp-intermediate-topics\TupleAliasInCSharp\tests> **dotnet test**

### CODE REVIEW

File: csharp-intermediate-topics/TupleAliasInCSharp/Tests/UnitTest1.cs
Comment (karimkod) [inline csharp-intermediate-topics/TupleAliasInCSharp/Tests/UnitTest1.cs:1]: This test project is empty, you can either add simple test to test the tuples, or just remove the project (which is better than having it empty or useless)

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/GlobalAlias/GlobalAlias.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿namespace TupleAliasInCSharp.GlobalAlias
```
Comment (karimkod) [inline csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/GlobalAlias/GlobalAlias.cs:1]: Let's use file-scoped namespaces.

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/Program.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+﻿global using tupleAliasGlobalUsing = (int num1, string str, int num2);
+global using System;
```
Comment (karimkod) [inline csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/Program.cs:2]: If this is really needed, let's move it to the Usings.cs as well. It's conventional to have all global usings in one file called Usings.cs

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAliasInCSharp.csproj [hunk 3]
```diff
@@ -0,0 +1,11 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net7.0</TargetFramework>
+    <DisableImplicitUsings>enable</DisableImplicitUsings>
+    <Nullable>enable</Nullable>
+    <LangVersion>preview</LangVersion>
```
Comment (karimkod) [inline csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAliasInCSharp.csproj:8]: With dotnet 7.0 installed on my computer I wasn't able to compile the code. And dotnet 7.0 is not a preview so I think the correct version of the framework should be 8.0 a not 7.0 and we should also point that to the readers, otherwise they will get confused, like me :)

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAlias.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
+﻿namespace TupleAliasInCSharp
```
Comment (karimkod) [inline csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAlias.cs:1]: We can use file-scoped namespaces here as well

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/Program.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+﻿global using tupleAliasGlobalUsing = (int num1, string str, int num2);
```
Comment (karimkod) [inline csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/Program.cs:1]: The tuple alias name should be UpperCamelCase, and we can move this into a separate file called Usings.cs

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAlias.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+{
+    using tupleAliasUsing = (int num1, string str, double num2);
```
Comment (karimkod) [inline csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAlias.cs:3]: The tuple alias should be UpperCamelCase.

**Review by karimkod — CHANGES_REQUESTED** [review]

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAlias.cs (hunk 6 above)
Comment (AditiSaxena0289) [reply csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAlias.cs]: @karimkod - This PR has been updated to reflect this change.

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/Program.cs (hunk 5 above)
Comment (AditiSaxena0289) [reply csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/Program.cs]: This PR has been updated to reflect recent changes.

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAlias.cs (hunk 4 above)
Comment (AditiSaxena0289) [reply csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAlias.cs]: This PR has been updated to reflect recent changes.

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAliasInCSharp.csproj (hunk 3 above)
Comment (AditiSaxena0289) [reply csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAliasInCSharp.csproj]: This PR has been updated to reflect this change.

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/Program.cs (hunk 2 above)
Comment (AditiSaxena0289) [reply csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/Program.cs]: This PR has been updated to reflect this change.

File: csharp-intermediate-topics/TupleAliasInCSharp/Tests/UnitTest1.cs
Comment (AditiSaxena0289) [reply csharp-intermediate-topics/TupleAliasInCSharp/Tests/UnitTest1.cs:1]: Test project is removed.

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/GlobalAlias/GlobalAlias.cs (hunk 1 above)
Comment (AditiSaxena0289) [reply csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/GlobalAlias/GlobalAlias.cs]: This PR has been updated to reflect this change.

**Comment by AditiSaxena0289** [comment]
@karimkod - I have made the revisions to the code as per your feedback and updated this PR. Please review them and let me know your comments.

Thank you.

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/GlobalAlias/GlobalAlias.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿namespace TupleAliasInCSharp.GlobalAlias;
+
+internal class GlobalAlias
+{
+    //ACCESS TUPLE WITH GLOBAL USING DIRECTIVE
```
Comment (jshergal) [inline csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/GlobalAlias/GlobalAlias.cs:5]: This comment is not needed

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAlias.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
+﻿namespace TupleAliasInCSharp;
+
+using EmployeeFinanaceDetails = (int Id, string FName, string LName, double salary);
+
+public class TupleAlias
+{
+    public static void PrintTupleValues()
+    {
+        //ACCESS TUPLE WITH GLOBAL USING DIRECTIVE
```
Comment (jshergal) [inline csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAlias.cs:9]: I am not sure the comments are really necessary here, but if we want to keep them (up to your editor @karimkod for sure), we should make them sentence case rather than all uppercase.

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAliasInCSharp.csproj [hunk 9]
```diff
@@ -0,0 +1,11 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net8.0</TargetFramework>
```
Comment (jshergal) [inline csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAliasInCSharp.csproj:5]: Until .NET 8.0 is officially released, we should target .NET 7.0

Comment (jshergal) [inline csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAliasInCSharp.csproj:5]: Ignore my previous comment. I added it before thinking about the fact that this code is using a new feature of .NET 8.0

**Comment by jshergal** [comment]
One other note. There is no unit test project along with this code. While this is a mini article and I understand it might be difficult to create some unit tests, under the CodeMaze guidelines, projects should contain unit tests. Again, final say up to your editor @karimkod and Sr. Review, but we should consider what kind of tests we could add to this.

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/GlobalAlias/GlobalAlias.cs [hunk 10]
```diff
@@ -0,0 +1,12 @@
+﻿namespace TupleAliasInCSharp.GlobalAlias;
+
+internal class GlobalAlias
```
Comment (karimkod) [inline csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/GlobalAlias/GlobalAlias.cs:3]: Let's change this to be "GlobalAliasUsage"

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAlias.cs [hunk 11]
```diff
@@ -0,0 +1,39 @@
+﻿namespace TupleAliasInCSharp;
+
+using EmployeeFinanaceDetails = (int Id, string FName, string LName, double salary);
```
Comment (karimkod) [inline csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAlias.cs:3]: typo

File: csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAlias.cs (hunk 8 above)
Comment (karimkod) [reply csharp-intermediate-topics/TupleAliasInCSharp/TupleAliasInCSharp/TupleAlias.cs:9]: @AditiSaxena0289 we don't need all these comments. the WriteLine already explains what's happening.

**Review by karimkod — CHANGES_REQUESTED** [review]
